### ISSUE TITLE
Overload resolution not picking first matching overload

### ISSUE THREAD

**Issue by Jeremiah-England** [astral-sh/ty#3076]
### Summary

[Playground](https://play.ty.dev/1081223e-8bb4-419d-a669-272618954e4b)

```python
from typing import TYPE_CHECKING, overload
from collections.abc import Iterable
import builtins

if TYPE_CHECKING:

    @overload
    def zip_typed[T1](iter1: Iterable[T1], /) -> builtins.zip[tuple[T1]]: ... 

    @overload
    def zip_typed[T1, T2](
        iter1: Iterable[T1], iter2: Iterable[T2], /, *, strict: bool
    ) -> builtins.zip[tuple[T1, T2]]: ...

    @overload
    def zip_typed[T](*iters: Iterable[T], strict: bool) -> builtins.zip[tuple[T, ...]]: ...

else:
    zip_typed = builtins.zip

x: list[int] = []
y: list[str] = []
z = zip_typed(x, y, strict=True)  # Unknown! Should be `zip[tuple[int, str]]`.
```

I think the inferred type here for `z` should be `zip[tuple[int, str]]`.

It appears that the variadic `*iters: ...` overload is not being eliminated at step 5 of the overload resolution algorithm due to a perfect match for the 2-arity overload: https://typing.python.org/en/latest/spec/overload.html#step-5.

Flipping the order of the variadic vs 2-arity overload does not change the outcome, so it seems like `ty` may not be picking the first matching overload in general? Uncommenting either of them gives the expected result.

### Version

ty 0.0.23

**Comment by carljm** [comment]
Thanks for the report! Definitely looks like a bug. It does seem like we are failing to eliminate the third overload for some reason, and so resolving to `Unknown` according to the ambiguous-overload-match rule in step 5. But it should be the case that all materializations of the arguments would match the second overload, so the third should be eliminated.

**Comment by dhruvmanila** [comment]
This definitely look like #1825
